src/txn/txn_ckpt.c

```diff
@@ -894,13 +917,22 @@ __wt_txn_checkpoint(WT_SESSION_IMPL *session, const char *cfg[])
 	 */
 	WT_RET(__wt_session_reset_cursors(session, false));
 
+	/* Ensure the metadata table is open before taking any locks. */
+	WT_RET(__wt_metadata_cursor(session, NULL));
+
 	/*
 	 * Don't highjack the session checkpoint thread for eviction.
 	 *
 	 * Application threads are not generally available for potentially slow
 	 * operations, but checkpoint does enough I/O it may be called upon to
 	 * perform slow operations for the block manager.
+	 *
+	 * Application checkpoints wait until the checkpoint lock is available,
+	 * compaction checkpoints don't.
 	 */
+#define	WT_TXN_SESSION_MASK						\
+	(WT_SESSION_CAN_WAIT | WT_SESSION_NO_EVICTION)
```

Comment:
> Can we get rid of `WT_SESSION_CAN_WAIT`? Is there any case where we know it helps with a workload?

It's a latency question, not a throughput question: application threads don't set `WT_SESSION_CAN_WAIT`, and so won't get highjacked by the block manager to call `fsync` if `os_cache_dirty_max` or `os_cache_max` is configured to `WT_SESSION::create`.

We could replace it with a test of `WT_SESSION_INTERNAL`, though, if there's no internal thread we're not willing to tap for an fsync call. (That change gets rid of `WT_THREAD_CAN_WAIT` too).

Let me know if you want me to make that change.